### ISSUE TITLE
stack needs to describe tarball dependency twice: in library and in application using this library

### ISSUE THREAD

**Issue by bapcyk** [commercialhaskell/stack#3742]
### General summary/comments (optional)

I have library package and application package. I want to use Cardinality package in my library package which is available only as tarball. So, I added to my library stack.yaml:

```
extra-deps:
- archive: https://hackage.haskell.org/package/Cardinality-0.2/Cardinality-0.2.tar.gz
```
and in library's cabal file:

```
build-depends: ...
               , Cardinality
```
Library package is building without problems. Now I want to build my application package which depends on my library package. Dependency is defined as:

```
packages:
  ...
  - '../xxx-common-utils'
```

in application's stack.yaml file. When I ran `stack build` in application - build fails (in library - it's fine). I found that if I include Cardinality dependency in application stack.yaml too - build passes. But this looks wrong: stack should build library package with Cardinality as dependency and then build application package without need to describe Cardinality in application stack.yaml too!

I tried different configurations: tarball in packages section, in extra-dep, with allow-newer: true, etc - results are different, but build always is broken: of library or of application. Build was passed only if I add tarball dep twice: in library and application.

### Steps to reproduce

1. Create 2 projects: application (simple) and library (simple-library)
2. Add Cardinality as dep. to library
3. Add library as dep. to application
4. Build library with stack - must be fine
5. Build application - must fail.

### Expected

Successful build with adding of Cardinality tarball dependency only once - in library package.

### Actual

Error:

```
Error: While constructing the build plan, the following exceptions were encountered:

In the dependencies for xxx-common-utils-0.2.0.0:
    Cardinality must match -any, but the stack configuration has no specified version (latest matching version is 0.2)
needed since xxx-common-utils is a build target.

Some potential ways to resolve this:

  * Recommended action: try adding the following to your extra-deps in /home/xxx/my-app/stack.yaml:

- Cardinality-0.2

  * Set 'allow-newer: true' to ignore all version constraints and build anyway.

  * You may also want to try using the 'stack solver' command.

Plan construction failed.
```

### Stack version

```
$ stack --version
Version 1.6.3, Git revision b27e629b8c4ce369e3b8273f04db193b060000db (5454 commits) x86_64 hpack-0.20.0
```
### Method of installation

Installed as OS package. Then upgraded several times.

**Comment by mgsloan** [comment]
This is just how stack works.  Only packages that are in the snapshot can be used as dependencies, everything else gets added to extra-deps.  If it arbitrarily decided to use the latest version of Cardinality from hackage, then builds could break as soon as it is updated.  Perhaps https://docs.haskellstack.org/en/stable/GUIDE/ will be helpful here.

Also, no need to use `archive:`.  You can just do

```yaml
extra-deps:
- Cardinality-0.2
```

**Comment by bapcyk** [comment]
Seems that it's bug in stack. Why I think so:
I can add Cardinality to my lib. OK, let's do it as `extra-deps: [Cardinality-0.2]`. All is fine, my lib build passes. No problem.
Next step: I want to compile my application (another Haskell project/package) which depends on my lib. package. And I can do it without to include Cardinality as dependency: but now in application project too! So, I need to include Cardinality as dependency twice: but I use it only in my lib. package, not in application! So, this is a bug.

**Comment by bapcyk** [comment]
What I see is:

```
In the dependencies for xxx-common-utils-0.2.0.0:
    Cardinality must match -any, but the stack configuration has no specified version (latest matching version is 0.2)
needed since xxx-common-utils is a build target.
```

This error is shown when I run `stack build` in application folder (where is located app's stack.yaml). So, stack understands that (you see this in error message):

* app depends on xxx-common-utils
* xxx-common-utils depends on Cardinality-0.2
* no such package in configuration blah-blah-blah
* app stack.yaml needs Cardinality as dependency too! Which sure is not right: we don't need to repeat all dependencies from using libraries (they should be defined once - in package of libraries)

This can happen if stack does not make separate compilation, for example, it INCLUDES lib dependency but does not compile it independently. But interesting that it does not happen with non-tarball dependencies.

@mgsloan : will be good to reopen bug or to find solution for such situation. I can not find relevant documentation because stack documentation (guide) seems superficial for this use-case

**Comment by bapcyk** [comment]
This is the stack.yaml of the library:

```
    resolver: lts-8.13
    packages:
    - '.'
    - '../xxx-test-utils'
    extra-deps:
      - Cardinality-0.2
    flags: {}
    extra-package-dbs: []
```

This - of application:

```
    resolver: lts-8.13
    packages:
    - '.'
    - '../xxx-common-utils'
    - '../xxx-atlassian-utils'
    - '../xxx-git-utils'
    - '../xxx-test-utils'
    extra-deps: []
    flags: {}
    extra-package-dbs: []
```

**Comment by mgsloan** [comment]
Yes, stack only cares about one stack.yaml at a time.  The library's configuration is specified by package.yaml or *.cabal.  "How to resolve dependencies" is specified by your stack.yaml.  If this redundancy bugs you, then just have a repository that contains both the library and the application, and have just one stack.yaml

> This can happen if stack does not make separate compilation, for example, it INCLUDES lib dependency but does not compile it independently. But interesting that it does not happen with non-tarball dependencies.

I don't quite follow how it is behaving differently with archives?

**Comment by bapcyk** [comment]
Hmmm, I found the cause of error!
When I add Cardinality in library's cabal file - this error happens! stack says that it can match/find "Cardinality" but can "Cardinality-0.2", right? So, question now is how to write "Cardinality-0.2" as "build-depends" line in lib's cabal file - because "-0.2" seems to be problem for cabal file parser, it says:

```
Unable to parse cabal file /home/yyy/xxx-common-utils.cabal: NoParse "build-depends" 26
```

**Comment by bapcyk** [comment]
I check .stack-work/ in lib and see that it was compiled and there are several `*Cardinality-02*` (wildcard) files there. So, stack (cabal?) can not make its work. I tried `stack solver` in the app folder and got:

```
Using configuration file: stack.yaml
Using cabal packages:
- /home/yyy/prj/xxx-atlassian-utils/
- /home/yyy/prj/xxx-common-utils/
- /home/yyy/prj/xxx-git-utils/
- /home/yyy/prj/xxx-test-utils/
- ./

Using resolver: lts-8.13
Using compiler: ghc-8.0.2
Asking cabal to calculate a build plan...
Trying with packages from lts-8.13 as hard constraints...
Attempt failed.

>>>> Cabal errors begin
cabal: Could not resolve dependencies:
trying: xxx-common-utils-0.2.0.0 (user goal)
trying: xxx-test-utils-0.1.0.0 (user goal)
Dependency tree exhaustively searched.
<<<< Cabal errors end

Retrying with packages from lts-8.13 as preferences...
Attempt failed.

>>>> Cabal errors begin
cabal: Could not resolve dependencies:
trying: xxx-common-utils-0.2.0.0 (user goal)
trying: xxx-test-utils-0.1.0.0 (user goal)
Dependency tree exhaustively searched.
<<<< Cabal errors end

*** Failed to arrive at a workable build plan.

Solver could not resolve package dependencies.
You can try the following:
    - Update external packages with 'stack update' and try again.
    - Tweak stack.yaml and try again
        - Remove any unnecessary packages.
        - Add any missing remote packages.
        - Add extra dependencies to guide solver.
        - Adjust resolver.
```
This happens with `extra-deps: [Cardinality-0.2]` in lib's stack.yaml and `, Cardinality` in lib's cabal file. So, looks like cabal can't find Cardinality, but it was compiled in lib's folder (as I said early) and exists in .stack-work/ caches.

**Comment by mgsloan** [comment]
I don't know why solver is failing.

In cabal, dependencies come with dependency ranges.  Like `cardinality >= 0.2`

**Comment by bapcyk** [comment]
I tried to add in lib's cabal file `, Cardinality >= 0.2` - the same error. But if I add `, cardinality >= 0.2` (lower-case 1st letter) - error is different, Cardinality module is not found at whole. So, problem may be in version mismatch? Do you know: may be problem is in version annotation (or package meta-info) of Cardinality package itself? Is it possible?

**Comment by bapcyk** [comment]
IMHO good is to distinguish source of error: Cardinality package, stack, cabal, my local stack/cabal files ?

**Comment by bapcyk** [comment]
@mgsloan I added attachment which shown the problem. There are 2 projects: mylib and myapp. So, try please:

```
cd mylib
stack build
# here all must be fine
cd ../myapp
stack build
# look to the error message now
```

[bug.tar.gz](https://github.com/commercialhaskell/stack/files/1603094/bug.tar.gz)

**Comment by mgsloan** [comment]
When using stack, you don't really need constraint bounds like `Cardinality >= 0.2`.

As I said before, you need `Cardinality-0.2` in both stack.yaml..

**Comment by bapcyk** [comment]
What I can't understand in this case: if myapp has dependecies in its extra-deps and they have own dependencies, why we should re-include all extra-dependencies of our extra-dependencies again?! Michael, IMHO it's bad user experience: from my point of view "extra-deps" don't looks like real dependencies but like "associated build job" or "textual including". Why we need to repeat part of dependencies graph?! Another question is: if I have local library and local application, how I should define dependency between them?! Because now I see that local path in "extra-deps:" or "packages:" is not real dependency. Can you imagine 100 local libraries with 5-10 extra-deps and 10 local applications? What to do in this case?! This is terrible user-experience... Sure, it's my IMHO but I never see such thing in other languages/pkg managers. IMHO will be better to have some way/workaround to solve such problem.

**Comment by bapcyk** [comment]
and 2nd interesting question is: if I define mylib/myapp dependency in "extra-deps:" section instead of "packages:" section (see attachment) - problem is solved! But in complex application/libraries with many dependencies this trick does not work. Which means for me that stack has different behavior in both cases with similar configuration (but different list of dependencies). This looks like bug too.

**Comment by mgsloan** [comment]
It is not a bug.  The package declares what it is compatible with.  The stack file declares how to build the project which consists of multiple packages.

The issue tracker is not a good place to answer questions that could easily be answered by reading manuals and googling.  Please learn a bit more about how things work before assuming the software is broken.

> Sure, it's my IMHO but I never see such thing in other languages/pkg managers

Which ones?

**Comment by mgsloan** [comment]
I think what you are misunderstanding is maybe that you don't need one stack file per package.  You just need one stack.yaml per "build environment" that can then have all of your 100 local libraries and 10 applications.

**Comment by bapcyk** [comment]
OK, but in this case no way to separate such local library and to publish it somewhere without to modify it (create new stack.yaml file). If I'll use one big stack.yaml file for all projects this means that I can't create projects from stack's template - now it is one project and so on. This looks very-very strange. You always can create library with `stack new simple-library` for example, right? And then you can decide to use it in some your project. So, how to do it correctly if you will need to repeat extra-deps of your library?! But library for me is like "black box" - one entity/one dependency.

In other language I can create normal library, build it and install it locally (in system or in "sandbox") or referent it. In C/C++ with Makefiles I can easy build separately real library and to define it as dependency of application instead of create one big Makefile per all my libraries and apps which use them. Your way (with one stack.yaml) in C/C++ world means: to combine all libraries sources and applications and to build it together instead of build separately and link app with libs. So, one devs group build libraries, another break their projects: extracts sources and inject them in another project? It sound terrible...

Michael, excuse my that I wrote my questions here because no other place to ask somebody: mail lists ignored this question, stackoverflow (except Michael Snoyman) too and no area where it can be discussed. I know that many projects uses bug trackers. for discussion in GitHub, so I decide to do it here :( Excuse me again.

**Comment by mgsloan** [comment]
That's ok. Just consider that your tone of saying things are a "terrible user experience", when you clearly do not yet understand them, is not very polite or conducive to constructive discussion.

> In C/C++ with Makefiles I can easy build separately real library and to define it as dependency of application instead of create one big Makefile per all my libraries and apps which use them.

C/C++ makefiles are a total nightmare compared to this.  Makefiles don't even usually specify dependency versions, you need to use autoconf or cmake or similar.  And then, you are in the same boat of every place you declare a build environment you need to specify all transitive dependencies.

So yeah, makefiles leave it up to the user to install the right dependencies which means easily shooting yourself in the foot.  It's a total nightmare compared to stack or even cabal-install, which allows specifying your dependency versions.

The only system I know of that makes it straightforward to do what you are asking for is nvm for node.  They traditionally did some absolutely crazy things, I hear that they are backpedaling a bit on this design decision.  The paradigm is that you can declare your library depends on a specific version of another library.  If two libraries depend on different versions of some other library, then that library gets copied.  This means that if the data type for that 3rd library has changed its format, things will break.  Problem is, JS doesn't have a type system by default so you only find out at runtime.

So, yeah.  You are either asking for stack to have non-reproducible builds and let you shoot yourself in the foot, or you're asking for some serious magic.  I have no idea how this magic would work, you haven't given a concrete implementation approach.

**Comment by bapcyk** [comment]
Michael, excuse me again, I really did not understand your point right: "packages:" describes packages OF THE CURRENT PROJECT and it is not for "linking" of "libraries". So, you are right. I found [interesting post of Michael Snoyman about this and he already answered about such feature](https://www.fpcomplete.com/blog/2017/07/stacks-new-extensible-snapshots?success=true).

So, we did not understand each other :) I'm talked about missing functionality but with wrong terminology, so the way how works "packages:" is not crazy-user-experience, but I'll wait with impatience such new feature. What about "magic" - IMHO if I define already "link" between 2 folders with "packages:" and both of them have own "stack.yaml" they can be treated as some kind of relation with direct (from one to second). So, if I have deps. in "extra-deps:" of "slave" project why stack can not clone them automatically in "extra-deps:" of the "master" project while it builds dependency graph? This will simplify developers life :)  Sure, I don't know interior of the stack, so I don't know: 1) how is it easy  2) whether it will contradict with some stack "ideas"

You are right about Makefile but tracking of versions is not a job of Makefile. I'm talking about absolutely another its feature: I can build hundreds libraries and to link them with hundreds of consumers (applications) of those libraries instead of combine sources of libraries and applications and build them together (NxM matrix). Stack currently offers to "include/inject" sources from library to application with "packages:" list in stack.yaml. Another options, as I understood, is to use one big stack.yaml per all projects which is not an option at whole. Sure, may be I'm missing something about stack and there are another workarounds :)
This was my point. Thanks for explanation!
